test/hotspot/jtreg/compiler/c2/irTests/InvolutionIdentityTests.java

```diff
@@ -0,0 +1,152 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package compiler.c2.irTests;
+
+import compiler.lib.ir_framework.DontCompile;
+import compiler.lib.ir_framework.IR;
+import compiler.lib.ir_framework.IRNode;
+import compiler.lib.ir_framework.Run;
+import compiler.lib.ir_framework.RunInfo;
+import compiler.lib.ir_framework.Test;
+import compiler.lib.ir_framework.TestFramework;
+import jdk.test.lib.Asserts;
+
+/*
+ * @test
+ * @bug 8350988
+ * @summary Test that Identity simplifications of Involution nodes are being performed as expected.
+ * @library /test/lib /
+ * @run driver compiler.c2.irTests.InvolutionIdentityTests
+ */
+public class InvolutionIdentityTests {
+    public static void main(String[] args) {
+        TestFramework.run();
+    }
+
+    @Run(test = {
+        "testI1", "testI2",
+        "testL1", "testL2",
+        "testF1",
+        "testD1"
+    })
+    public void runMethod() {
+        int ai = RunInfo.getRandom().nextInt();
+
+        int mini = Integer.MIN_VALUE;
+        int maxi = Integer.MAX_VALUE;
+
+        assertResultI(0);
+        assertResultI(ai);
+        assertResultI(mini);
+        assertResultI(maxi);
+
+        long al = RunInfo.getRandom().nextLong();
+
+        long minl = Long.MIN_VALUE;
+        long maxl = Long.MAX_VALUE;
+
+        assertResultL(0);
+        assertResultL(al);
+        assertResultL(minl);
+        assertResultL(maxl);
+
+        float af = RunInfo.getRandom().nextFloat();
+        float inf = Float.POSITIVE_INFINITY;
+        float nanf = Float.NaN;
+
+        assertResultF(0f);
+        assertResultF(-0f);
+        assertResultF(af);
+        assertResultF(inf);
+        assertResultF(nanf);
+
+        double ad = RunInfo.getRandom().nextDouble();
+        double ind = Double.POSITIVE_INFINITY;
+        double nand = Double.NaN;
+
+        assertResultD(0d);
+        assertResultD(-0d);
+        assertResultD(ad);
+        assertResultD(ind);
+        assertResultD(nand);
+
+    }
+
+    @DontCompile
+    public void assertResultI(int a) {
+        Asserts.assertEQ(Integer.reverseBytes(Integer.reverseBytes(a)), testI1(a));
+        Asserts.assertEQ(Integer.reverse(Integer.reverse(a))          , testI2(a));
+    }
+
+    @DontCompile
+    public void assertResultL(long a) {
+        Asserts.assertEQ(Long.reverseBytes(Long.reverseBytes(a)), testL1(a));
+        Asserts.assertEQ(Long.reverse(Long.reverse(a))          , testL2(a));
+    }
+
+    @DontCompile
+    public void assertResultF(float a) {
+        Asserts.assertEQ(Float.floatToRawIntBits(-(-a)), Float.floatToRawIntBits(testF1(a)));
+    }
+
+    @DontCompile
+    public void assertResultD(double a) {
+        Asserts.assertEQ(Double.doubleToRawLongBits(-(-a)), Double.doubleToRawLongBits(testD1(a)));
+    }
+
+    @Test
+    @IR(failOn = {IRNode.REVERSE_BYTES_I})
```

Comment:
I'm not sure if this is fine as the ReverseBytes nodes depend on intrinsics. From my understanding, the methods are just seen as normal methods on platforms without reverseBytes support. In that case, the test would still pass, but it might be surprising that it passes. Is this fine or is there a better way here?

Reply:
Interesting question - expanding on that, could arbitrary methods be marked as self-inverse to be represented in the IR?

Reply:
I don't think there are many such methods, and my knowledge of Ideal isn't good enough to judge whether that's possible.

But it might make sense to use the nodes even when the intrinsic isn't available, and "lower" it to the existing implementation (either a call or inlining) after optimizing the node itself.